File: app/src/main/java/com/github/sewerina/meter_readings/ui/report/ReportViewModel.java
package com.github.sewerina.meter_readings.ui.report;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.github.sewerina.meter_readings.database.AppDao;
import com.github.sewerina.meter_readings.database.HomeEntity;
import com.github.sewerina.meter_readings.database.ReadingEntity;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

import javax.inject.Inject;
import javax.inject.Named;

import io.reactivex.Completable;
import io.reactivex.CompletableEmitter;
import io.reactivex.CompletableOnSubscribe;
import io.reactivex.CompletableSource;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

public class ReportViewModel extends ViewModel {
    private final CompositeDisposable mDisposables = new CompositeDisposable();
    private MutableLiveData<List<Report>> mReports = new MutableLiveData<>();

    @Inject
    AppDao mDao;

    @Inject
    @Named("reports")
    CollectionReference mReportCollectionReference;

    public LiveData<List<Report>> getReports() {
        return mReports;
    }

    public void addReport(HomeEntity currentHomeEntity) {
        Disposable subscribe = mDao.getReadingsForHomeRx(currentHomeEntity.id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .flatMap(new Function<List<ReadingEntity>, SingleSource<Report>>() {
                    @Override
                    public SingleSource<Report> apply(final List<ReadingEntity> readingEntities) throws Exception {
                        return Single.fromCallable(new Callable<Report>() {
                            @Override
                            public Report call() throws Exception {
                                Report report = new Report(readingEntities);
                                List<Report> reportList = mReports.getValue();

                                if (reportList == null) {
                                    reportList = new ArrayList<>();
                                }
                                reportList.add(report);
                                mReports.postValue(reportList);

                                return report;
                            }
                        });
                    }
                })
                .flatMapCompletable(new Function<Report, CompletableSource>() {
                    @Override
                    public CompletableSource apply(Report report) throws Exception {
                        return addReportInCloudFirestore(report);
                    }
                })
                .subscribe();
        mDisposables.add(subscribe);
    }

    private Completable addReportInCloudFirestore(final Report report) {
        return Completable.create(new CompletableOnSubscribe() {
            @Override
            public void subscribe(final CompletableEmitter emitter) throws Exception {
                mReportCollectionReference
                        .add(report)
                        .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
                            @Override
                            public void onSuccess(DocumentReference documentReference) {
                                emitter.onComplete();
                            }
                        })
                        .addOnFailureListener(new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception e) {
                                emitter.onError(new Exception());
                            }
                        });
            }
        });
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        mDisposables.dispose();
    }
}

File: app/src/main/java/com/github/sewerina/meter_readings/ui/report/ReportViewModel.kt
package com.github.sewerina.meter_readings.ui.report

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.github.sewerina.meter_readings.R
import com.github.sewerina.meter_readings.database.AppDao
import com.github.sewerina.meter_readings.database.HomeEntity
import com.github.sewerina.meter_readings.ui.MessageService
import com.google.firebase.firestore.CollectionReference
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import javax.inject.Inject
import javax.inject.Named

@HiltViewModel
class ReportViewModel @Inject constructor(
    private val mDao: AppDao,

    @Named("reports")
    private val mReportCollectionReference: CollectionReference,

    private val mMessageService: MessageService
) : ViewModel() {
    private val mReports = MutableLiveData<MutableList<Report>>()
    val reports: LiveData<MutableList<Report>>
        get() = mReports

    fun addReport(currentHomeEntity: HomeEntity) {
        viewModelScope.launch {
            val readingEntities = mDao.getReadingsForHome(currentHomeEntity.id)
            if (readingEntities.isEmpty()) {
                mMessageService.showMessage(R.string.unable_generate_report)
                return@launch
            }

            val report = Report(readingEntities)
            val reportList = mReports.value ?: ArrayList()
            reportList.add(report)
            mReports.postValue(reportList)
            addReportInCloudFirestore(report)
        }
    }

    private suspend fun addReportInCloudFirestore(report: Report) {
        mReportCollectionReference
            .add(report)
            .await()
    }

    fun loadReports(currentHome: HomeEntity) {
        // 1. Пойти в Firebase & взять оттуда все reports для currentHome.Id
        // 2. Обновить этими reports MutableLiveData
        viewModelScope.launch {
            val reports = loadReportsFromCloudFirestore(currentHome.id)
            mReports.postValue(reports)
        }
    }

    private suspend fun loadReportsFromCloudFirestore(currentHomeId: Int): MutableList<Report> {
        val documents = mReportCollectionReference
            .whereEqualTo("homeId", currentHomeId)
            .get()
            .await()

        val reports: MutableList<Report> = ArrayList()
        for (document in documents) {
            val report = document.toObject(Report::class.java)
            reports.add(report)
        }
        return reports
    }
}